python-avd/pyavd/_schema/models/avd_model.py

```diff
@@ -79,19 +84,34 @@ def _from_dict(cls: type[T_AvdModel], data: Mapping, keep_extra_keys: bool = Fal
                 msg = f"Invalid key '{key}'. Not available on '{cls.__name__}'."
                 raise KeyError(msg)
 
-            cls_args[field] = coerce_type(data[key], cls._fields[field]["type"])
+            # Handle Dynamic Keys slightly differently
+            child_data_source = data_source if cls.__name__.startswith("Dynamic") and key == "value" else data_source.create_descendant(key)
+
+            cls_args[field] = coerce_type(data[key], cls._fields[field]["type"], data_source=child_data_source)
+            cls._field_source[field] = child_data_source
 
         if custom_data:
             cls_args["_custom_data"] = custom_data
 
-        return cls(**cls_args)
+        cls_instance = cls(**cls_args)
+        cls_instance._source = data_source
```

Comment:
The _source could be set in the same way as `_custom_data` so you can just add it to the cls_args.